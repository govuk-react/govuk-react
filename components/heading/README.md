Heading
=======

### Import
```js
  import Heading from '@govuk-react/heading';
```
<!-- STORY -->

### Usage


Simple

```jsx
import { Heading } from 'govuk-react'

<Heading>Heading text</Heading>
```

To pick different heading levels it is recommended to use the shortcut versions as
that will pick the appropriate tag as well as set the appropriate corresponding font size.

Using shortcuts

```jsx
import { H1, H2, H3, H4, H5, H6 } from 'govuk-react'

<H1>h1</H1>
<H2>h2</H2>
<H3>h3</H3>
<H4>h4</H4>
<H5>h5</H5>
<H6>h6</H6>
```

Differing sizes

```jsx
<H6 size={80}>
  H6 with font size 80
</H6>
<Heading as="h2" size="SMALL">
  Heading as h2 with SMALL size
</Heading>
<H3 size="LARGE">H3 with LARGE size</H3>
```

Props pass through

```jsx
<Heading onClick={() => { console.log('clicked'); }}>Click me</Heading>
```

### References:

- https://design-system.service.gov.uk/styles/typography/#headings
- https://github.com/alphagov/govuk_frontend_toolkit/blob/master/stylesheets/_typography.scss
- https://github.com/alphagov/govuk-frontend/blob/main/src/govuk/core/_typography.scss

### Properties
Prop | Required | Default | Type | Description
:--- | :------- | :------ | :--- | :----------
 `as` |  | ```undefined``` |  | Semantic heading tag to use (e.g. 'h3')<br/>By default element used will be an 'h1'
 `level` |  | ```undefined``` |  | Semantic heading level value between 1 and 6 (deprecated)
 `size` |  | ```'XLARGE'``` |  | Visual size level, accepts:<br/>   `XLARGE`, `LARGE`, `MEDIUM`, `SMALL`, `XL`, `L`, `M`, `S`<br/>   or a numeric size that fits in the GDS font scale list


