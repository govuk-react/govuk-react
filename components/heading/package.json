{
  "name": "@govuk-react/heading",
  "version": "0.9.1",
  "dependencies": {
    "@govuk-react/constants": "^0.9.1",
    "@govuk-react/lib": "^0.9.1"
  },
  "peerDependencies": {
    "react": ">=16.2.0",
    "styled-components": ">=5.1"
  },
  "scripts": {
    "docs": "doc-component ./lib/heading.js ./README.md"
  },
  "main": "dist/govuk-react-heading.cjs.js",
  "module": "dist/govuk-react-heading.esm.js",
  "source": "src/index.tsx",
  "author": "Alasdair McLeay",
  "license": "MIT",
  "homepage": "https://github.com/govuk-react/govuk-react/tree/main/components/heading",
  "description": "Outputs a styled heading tag of provided level, rendering children as contents.",
  "private": false,
  "publishConfig": {
    "access": "public"
  }
}
