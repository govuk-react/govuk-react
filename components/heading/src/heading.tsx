import type { StyledComponentProps } from 'styled-components';

import type { WithWhiteSpaceProps } from '@govuk-react/lib';

import styled from 'styled-components';
import React from 'react';
import { HEADING_SIZES, LEVEL_SIZE, LEVEL_TAG, MEDIA_QUERIES, TYPOGRAPHY_SCALE } from '@govuk-react/constants';
import { spacing, typography } from '@govuk-react/lib';

// use `size` only with string for XLARGE, SMALL etc and number for px size
// so if `size` is a string, we find a numeric size based off `HEADING_SIZES`
// but if `size` is a number we just send through that number

const StyledHeading = styled('h1')<StyledHeadingOwnProps>(
  typography.textColour,
  ({ size }) => {
    const actualSize = Number.isNaN(Number(size)) ? HEADING_SIZES[size] : size;

    if (!actualSize) {
      throw Error(`Unknown size ${size} used for heading.`);
    }

    return { ...typography.font({ size: actualSize, weight: 'bold' }) };
  },
  {
    display: 'block',
    marginTop: 0,
  },
  ({ size }) => {
    const actualSize = Number.isNaN(Number(size)) ? HEADING_SIZES[size] : size;
    const scaleInfo = TYPOGRAPHY_SCALE[actualSize];

    return {
      marginBottom: scaleInfo.mobile.spacing,
      [MEDIA_QUERIES.TABLET]: {
        marginBottom: scaleInfo.tablet.spacing,
      },
    };
  },
  spacing.withWhiteSpace()
);

/**
 *
 * ### Usage
 *
 *
 * Simple
 *
 * ```jsx
 * import { Heading } from 'govuk-react'
 *
 * <Heading>Heading text</Heading>
 * ```
 *
 * To pick different heading levels it is recommended to use the shortcut versions as
 * that will pick the appropriate tag as well as set the appropriate corresponding font size.
 *
 * Using shortcuts
 *
 * ```jsx
 * import { H1, H2, H3, H4, H5, H6 } from 'govuk-react'
 *
 * <H1>h1</H1>
 * <H2>h2</H2>
 * <H3>h3</H3>
 * <H4>h4</H4>
 * <H5>h5</H5>
 * <H6>h6</H6>
 * ```
 *
 * Differing sizes
 *
 * ```jsx
 * <H6 size={80}>
 *   H6 with font size 80
 * </H6>
 * <Heading as="h2" size="SMALL">
 *   Heading as h2 with SMALL size
 * </Heading>
 * <H3 size="LARGE">H3 with LARGE size</H3>
 * ```
 *
 * Props pass through
 *
 * ```jsx
 * <Heading onClick={() => { console.log('clicked'); }}>Click me</Heading>
 * ```
 *
 * ### References:
 *
 * - https://design-system.service.gov.uk/styles/typography/#headings
 * - https://github.com/alphagov/govuk_frontend_toolkit/blob/master/stylesheets/_typography.scss
 * - https://github.com/alphagov/govuk-frontend/blob/main/src/govuk/core/_typography.scss
 */
export const Heading: HeadingType = ({ level = undefined, ...props }: React.ComponentProps<HeadingType>) => {
  if (level) {
    if (process.env.NODE_ENV !== 'production') {
      // eslint-disable-next-line no-console
      console.warn(`deprecated prop 'level' used in Heading, please replace with an "H${level}" component`);
    }
    if (LEVEL_TAG[level]) {
      return <StyledHeading size={LEVEL_SIZE[level]} {...props} as={LEVEL_TAG[level]} />;
    }
  }

  return <StyledHeading {...props} />;
};

Heading.defaultProps = {
  level: undefined,
  size: 'XLARGE',
};

Heading.displayName = 'Heading';

interface HeadingType extends React.FC<HeadingOwnProps> {
  (props: HeadingPropsWithoutAs): React.ReactElement<HeadingPropsWithoutAs>;
  <AsC extends string | React.ComponentType = 'h1', FAsC extends string | React.ComponentType = AsC>(
    props: HeadingPropsWithAs<AsC, FAsC>
  ): React.ReactElement<HeadingPropsWithAs<AsC, FAsC>>;
}

type HeadingPropsWithoutAs = StyledComponentProps<'h1', never, HeadingOwnProps, never> & {
  as?: never | undefined;
  forwardedAs?: never | undefined;
};

type HeadingPropsWithAs<AsC extends string | React.ComponentType, FAsC extends string | React.ComponentType = AsC> =
  StyledComponentProps<AsC, never, HeadingOwnProps, never, FAsC> & {
    as?: AsC | undefined;
    forwardedAs?: FAsC | undefined;
  };

interface StyledHeadingOwnProps extends WithWhiteSpaceProps {
  /**
   * Visual size level, accepts:
   *    `XLARGE`, `LARGE`, `MEDIUM`, `SMALL`, `XL`, `L`, `M`, `S`
   *    or a numeric size that fits in the GDS font scale list
   */
  size?: number | string;
}

interface HeadingOwnProps {
  /**
   * Semantic heading level value between 1 and 6 (deprecated)
   */
  level?: number | string;
  /**
   * Visual size level, accepts:
   *    `XLARGE`, `LARGE`, `MEDIUM`, `SMALL`, `XL`, `L`, `M`, `S`
   *    or a numeric size that fits in the GDS font scale list
   */
  size?: number | string;
}

export default Heading;
