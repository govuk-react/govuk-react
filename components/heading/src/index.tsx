import Heading from './heading';

export default Heading;
export { H1, H2, H3, H4, H5, H6 } from './presets';
