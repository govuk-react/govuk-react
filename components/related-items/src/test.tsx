import React from 'react';
import ReactDOM from 'react-dom';
import { mount } from 'enzyme';
import { H3 } from '@govuk-react/heading';
import UnorderedList from '@govuk-react/unordered-list';
import Link from '@govuk-react/link';
import ListItem from '@govuk-react/list-item';

import RelatedItems from '.';

const wrapper = (
  <RelatedItems>
    <H3>Travel abroad</H3>
    <UnorderedList listStyleType="none">
      <ListItem>
        <Link href="https://example.com">Link A</Link>
      </ListItem>
      <ListItem>
        <Link href="https://example.com">Link B</Link>
      </ListItem>
      <ListItem>
        <Link href="https://example.com">
          <strong>more</strong>
        </Link>
      </ListItem>
    </UnorderedList>
  </RelatedItems>
);

describe('RelatedItems', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(wrapper, div);
  });

  it('matches wrapper snapshot', () => {
    expect(mount(wrapper)).toMatchSnapshot('wrapper mount');
  });
});
