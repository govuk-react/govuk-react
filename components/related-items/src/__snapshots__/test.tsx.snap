// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelatedItems matches wrapper snapshot: wrapper mount 1`] = `
.c1 {
  color: #0b0c0c;
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 700;
  font-size: 18px;
  line-height: 1.1111111111111112;
  display: block;
  margin-top: 0;
  margin-bottom: 15px;
}

.c5 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 16px;
  line-height: 1.25;
  margin-bottom: 5px;
}

.c3 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 16px;
  line-height: 1.25;
  color: #0b0c0c;
  margin-top: 0;
  margin-bottom: 15px;
  list-style-type: none;
  padding-left: 0;
}

.c2 .c2 {
  margin-top: 10px;
}

.c3 > .c4 {
  margin-bottom: 5px;
}

.c6 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

.c6:focus {
  outline: 3px solid #ffdd00;
  outline-offset: 0;
  background-color: #ffdd00;
}

.c6:link {
  color: #1d70b8;
}

.c6:visited {
  color: #4c2c92;
}

.c6:hover {
  color: #2b8cc4;
}

.c6:active {
  color: #2b8cc4;
}

.c6:focus {
  color: #0b0c0c;
}

.c0 {
  border-top: 10px solid #1d70b8;
  padding-top: 5px;
  width: 100%;
  margin-bottom: 0;
}

.c0 > h3 {
  margin-bottom: 10px;
}

.c0 > ul {
  margin-bottom: 20px;
}

.c0 > ul > li {
  margin-bottom: calc(10px + 2px);
}

@media print {
  .c1 {
    color: #000;
  }
}

@media print {
  .c1 {
    font-size: 18px;
    line-height: 1.15;
  }
}

@media only screen and (min-width:641px) {
  .c1 {
    font-size: 24px;
    line-height: 1.25;
  }
}

@media only screen and (min-width:641px) {
  .c1 {
    margin-bottom: 20px;
  }
}

@media print {
  .c5 {
    font-size: 14px;
    line-height: 1.15;
  }
}

@media only screen and (min-width:641px) {
  .c5 {
    font-size: 19px;
    line-height: 1.3157894736842106;
  }
}

@media print {
  .c3 {
    font-size: 14px;
    line-height: 1.15;
  }
}

@media only screen and (min-width:641px) {
  .c3 {
    font-size: 19px;
    line-height: 1.3157894736842106;
  }
}

@media print {
  .c3 {
    color: #000;
  }
}

@media only screen and (min-width:641px) {
  .c3 {
    margin-bottom: 20px;
  }
}

@media print {
  .c6 {
    font-family: sans-serif;
  }
}

@media print {
  .c6[href^="/"]::after,
  .c6[href^="http://"]::after,
  .c6[href^="https://"]::after {
    content: " (" attr(href) ")";
    font-size: 90%;
    word-wrap: break-word;
  }
}

@media only screen and (min-width:641px) {
  .c0 {
    font-size: 16px;
    line-height: 1.25;
  }
}

@media only screen and (min-width:641px) {
  .c0 {
    margin-bottom: 0;
  }
}

<RelatedItems>
  <div
    className="c0"
  >
    <H3>
      <Heading
        as="h3"
        size="MEDIUM"
      >
        <heading__StyledHeading
          as="h3"
          size="MEDIUM"
        >
          <h3
            className="c1"
            size="MEDIUM"
          >
            Travel abroad
          </h3>
        </heading__StyledHeading>
      </Heading>
    </H3>
    <UnorderedList
      listStyleType="none"
    >
      <OrderedList
        as="ul"
        listStyleType="none"
      >
        <ul
          className="c2 c3"
        >
          <ListItem>
            <li
              className="c4 c5"
            >
              <Link
                href="https://example.com"
                muted={false}
                noVisitedState={false}
                textColour={false}
              >
                <a
                  className="c6"
                  href="https://example.com"
                >
                  Link A
                </a>
              </Link>
            </li>
          </ListItem>
          <ListItem>
            <li
              className="c4 c5"
            >
              <Link
                href="https://example.com"
                muted={false}
                noVisitedState={false}
                textColour={false}
              >
                <a
                  className="c6"
                  href="https://example.com"
                >
                  Link B
                </a>
              </Link>
            </li>
          </ListItem>
          <ListItem>
            <li
              className="c4 c5"
            >
              <Link
                href="https://example.com"
                muted={false}
                noVisitedState={false}
                textColour={false}
              >
                <a
                  className="c6"
                  href="https://example.com"
                >
                  <strong>
                    more
                  </strong>
                </a>
              </Link>
            </li>
          </ListItem>
        </ul>
      </OrderedList>
    </UnorderedList>
  </div>
</RelatedItems>
`;
